

https://cole-trapnell-lab.github.io/monocle3/docs/introduction/

# Install

```{r}
# BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
#                        'limma', 'lme4', 'S4Vectors', 'SingleCellExperiment',
#                        'SummarizedExperiment', 'batchelor', 'HDF5Array',
#                        'terra', 'ggrastr'))
# devtools::install_github('cole-trapnell-lab/monocle3')
library(monocle3)
```

# Getting started w/ Monocle 3
https://cole-trapnell-lab.github.io/monocle3/docs/getting_started/

## Get started

```{r Load Monocle3}
library(monocle3)

# The tutorial shown below and on subsequent pages uses two additional packages:
library(ggplot2)
library(dplyr)
```

Input: cell by gene expression matrix
- absolute transcript counts (from UMI experiments)
- works "out-of-the-box" w/ transcript count matrices produced by Cell Ranger

## `cell_data_set` class
- derived from Bioconductor `SingleCellExperiment` class
- three input files
  - `expression_matrix`: numeric mtx of expression values, rows are genes, columns are cells
    - ncol(exp_mtx) == nrow(cell_metadata)
    - nrow(exp_mtx) == nrow(gene_metadata)
    - colnames(exp_mtx) == rownames(cell_metadata)
    - rownames(exp_mtx) == rownames(gene_metadata)
  - `cell_metadata`: dataframe, rows are cells, columns are cell attributes (cell type, culture condition, day captured)
  - `gene_metadata`: dataframe, rows are features (genes), columns are gene attributes (biotype, gc content, etc.)
    - "gene_short_name_ column, which represents gene symbol or simple name (generally used for plotting) for each gene

```{r Generate cell_data_set}
# Load the data
expression_matrix <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_expression.rds"))
cell_metadata <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_colData.rds"))
gene_annotation <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_rowData.rds"))


# Make the CDS object
cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```

`umi_cutoff` set to 100 by default!
```{r Generate cell_data_set from 10X output, eval=FALSE}
# Provide the path to the Cell Ranger output.
cds <- load_cellranger_data("~/Downloads/10x_data")
# MatrixMarket format
cds <- load_mm_data(mat_path = "~/Downloads/matrix.mtx", 
                    feature_anno_path = "~/Downloads/features.tsv", 
                    cell_anno_path = "~/Downloads/barcodes.tsv")
```

```{r Sparse matrix for cell_data_set, eval=FALSE}
cds <- new_cell_data_set(as(umi_matrix, "sparseMatrix"),
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_metadata)
```

```{r Combine CDS objects, eval=FALSE}
# make a fake second cds object for demonstration
cds2 <- cds[1:100,]

big_cds <- combine_cds(list(cds, cds2))
```

# Clustering & classifying your cells

Cao & Packer et al. C. elegans data
```{r Load data}
library(monocle3)
library(dplyr) # imported for some downstream data manipulation

expression_matrix <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_expression.rds"))
cell_metadata <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_colData.rds"))
gene_annotation <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/cao_l2_rowData.rds"))

cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```

```{r Pre-process data}
# PCA is standard for RNA-seq
cds <- preprocess_cds(cds, num_dim = 100)
cds
plot_pc_variance_explained(cds)
```
`reduce_dimension`: UMAP by default
- faster + better suited for RNA-seq clustering + trajectory analysis
- can set `umap.fast_sgd=TRUE` to use fast stochastic gradient descent method inside of UMAP
- can also set `cores` arg for multithreading
- either options will make UMAP not replicable btwn runs

```{r Reduce dimensionality}
# UMAP by default 
cds <- reduce_dimension(cds)
# tSNE - won't overwrite
cds <- reduce_dimension(cds, reduction_method="tSNE")
```

`color_cell_by` can be any column in `colData(cds)`
```{r Visualize cells}
# UMAP Unannotated
plot_cells(cds) 
# UMAP Manual cell type annotation
plot_cells(cds, color_cells_by="cao_cell_type") 
# UMAP Gene expression
plot_cells(cds, genes=c("cpna-2", "egl-21", "ram-2", "inos-1")) 
# tSNE reduction
plot_cells(cds, reduction_method="tSNE", color_cells_by="cao_cell_type")
# PCA
plot_cells(cds, reduction_method="PCA", color_cells_by="cao_cell_type") 
```

Cao & Packer et al. "plate" annotation
- no evident batch effects
- can use `align_cds()` function to remove batch effect
  - continuous effects: uses Limma
  - discrete effects: uses Batchelor::fastMNN
```{r Batch effects}
plot_cells(cds, color_cells_by="plate", label_cell_groups=FALSE)
cds <- align_cds(cds, num_dim = 100, alignment_group = "plate")
cds <- reduce_dimension(cds)
plot_cells(cds, color_cells_by="plate", label_cell_groups=FALSE)
```

- Community detection (Levine et al. phenoGraph algorithm) to group cells into clusters
- Cells divided into partitions using statistical test from Alex Wolf et al. PAGA algorithm
```{r Cluster cells}
cds <- cluster_cells(cds, resolution=1e-5)
plot_cells(cds)
plot_cells(cds, color_cells_by="partition", group_cells_by="partition")
```

```{r Cluster marker genes}
marker_test_res <- top_markers(cds, group_cells_by="partition", 
                               reference_cells=1000, cores=8)
# Rank table according to specificity metrics, take top gene per cluster
# pseudo_R2
top_specific_markers <- marker_test_res %>%
  filter(fraction_expressing >= 0.10) %>%
  group_by(cell_group) %>%
  top_n(3, pseudo_R2)

top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))

# Plot expression + fraction of cells that express each marker in each group
plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="partition",
                    ordering_type="maximal_on_diag",
                    max.size=3)
```

# Constructing single-cell trajectories

Packer & Zhu et al. C. elegans time series embryo development scRNA data
```{r Load Packer & Zhu data}
expression_matrix <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_expression.rds"))
cell_metadata <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_colData.rds"))
gene_annotation <- readRDS(url("https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_rowData.rds"))

cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```

```{r Preprocess data}
# Batch correction
cds <- preprocess_cds(cds, num_dim = 50)
cds <- align_cds(
  cds,
  alignment_group = "batch",
  residual_model_formula_str = "~ bg.300.loading + bg.400.loading + bg.500.1.loading + bg.500.2.loading + bg.r17.loading + bg.b01.loading + bg.b02.loading")
```

```{r Reduce dimensionality + visualize}
cds <- reduce_dimension(cds)
plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "cell.type")

# View interesting genes
ciliated_genes <- c("che-1",
                    "hlh-17",
                    "nhr-6",
                    "dmd-6",
                    "ceh-36",
                    "ham-1")

plot_cells(cds,
           genes=ciliated_genes,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)
```

- Monocle doesn't assume all cells in dataset descend from common transcriptional "ancestor"
learns when cells should be placed in same trajectory vs separate trajectories through clustering procedure
- each partition will eventually become separate trajectory
```{r Cluster cells}
cds <- cluster_cells(cds)
plot_cells(cds, color_cells_by = "partition")
```

```{r Learn trajectory graph}
cds <- learn_graph(cds)
plot_cells(cds,
           color_cells_by = "cell.type",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)
```

Pseudotime: measure of how much progress an individual cell has made through a process such as cell differentiation
- abstract unit of progress
- distance btwn cell and the start of the trajectory, measured along the shortest path
- total length defined in terms of total amt of transcriptional change that cell undergoes as it moves from the starting state to the end state

Visualization
- black lines: structure of graph, not fully connected, distinct partitions
- light gray circles: leaf, different outcome/cell fate of trajectory
- black circles: branch nodes, which cells can travel through to outcome
```{r Order cells in pseudotime}
plot_cells(cds,
           color_cells_by = "embryo.time.bin",
           label_cell_groups=FALSE,
           label_leaves=TRUE,
           label_branch_points=TRUE,
           graph_label_size=1.5)
```

- gray cells: infinite pseudotime, not reachable from root nodes
  - any cell on partition w/o root node will be inf. pseudotime
  - in general, choose at least one root per partition
```{r Choose root + plot}
cds <- order_cells(cds)

plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5)
```

Ex. root node picker helper function
1. group cells according to which trajectory graph node they're nearest to
2. calculate what fraction of cells at each node come from earliest time point
3. pick node most heavily occupied by early cells, return as root
```{r Specify root node programatically}
# a helper function to identify the root principal points:
get_earliest_principal_node <- function(cds, time_bin="130-170"){
  cell_ids <- which(colData(cds)[, "embryo.time.bin"] == time_bin)
  
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}

cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))

plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5)
```

```{r Subset cells by branch}
cds_sub <- choose_graph_segments(cds) # Interactive
```


```{r 3D trajectories}
cds_3d <- reduce_dimension(cds, max_components = 3)
cds_3d <- cluster_cells(cds_3d)
cds_3d <- learn_graph(cds_3d)
cds_3d <- order_cells(cds_3d, root_pr_nodes=get_earliest_principal_node(cds))

cds_3d_plot_obj <- plot_cells_3d(cds_3d, color_cells_by="partition")

cds_3d_plot_obj
```






