
# CONICSmat fTutorial
Dataset: SmartSeq2 scRNA seq of Oligodendroglioma

https://github.com/diazlab/CONICS/wiki/Tutorial---CONICSmat;---Dataset:-SmartSeq2-scRNA-seq-of-Oligodendroglioma

## Setup

```{r Libraries}
install.packages("beanplot")
install.packages("mixtools")
install.packages("pheatmap")
install.packages("zoo")
install.packages("squash")
BiocManager::install("biomaRt")

# tSNE viz
install.packages("Rtsne")
BiocManager::install("scran")

devtools::install_github("diazlab/CONICS/CONICSmat", dep = FALSE)
library(CONICSmat)
```

```{r Load exp}
suva_expr = as.matrix(
  Rubrary::rread("/Users/liaoyj/Library/CloudStorage/Dropbox/OV/scRNA_UCLA/CONICSmat/OG_processed_data_portal.txt", row.names = 1))
suva_expr[which(is.na(suva_expr))]=0
dim(suva_expr)
Rubrary::corner(suva_expr, 5)
```

```{r Patient info}
patients=unlist(strsplit(colnames(suva_expr),"_",fixed=TRUE))[seq(1,(3*ncol(suva_expr))-1,3)]
unique(patients)

patients[which(patients=="X93")]="MGH93"
patients[which(patients=="X97")]="MGH97"
```

```{r Chr arm coordinates}
regions=read.table("/Users/liaoyj/Library/CloudStorage/Dropbox/OV/scRNA_UCLA/CONICSmat/chromosome_arm_positions_grch38.txt",sep="\t",row.names = 1,header = T)
head(regions,n=5)
```

## Sample dataset: Processing

```{r process}
# Chromosomal position of genes in exp mtx
gene_pos=getGenePositions(rownames(suva_expr), ensembl_version = "https://www.ensembl.org")
# Filter uninformative genes
suva_expr=filterMatrix(suva_expr,gene_pos[,"hgnc_symbol"],minCells=5)
# Normalization factor per cell
normFactor=calcNormFactors(suva_expr)
# Confirm bimodal distribution of avg gene expression
l=plotAll(suva_expr,normFactor,regions,gene_pos,"SUVA_CNVs")
# Heatmap of posterior probabilities of cells for comp2 of each region
hi=plotHistogram(l,suva_expr,clusters=2,zscoreThreshold=4,patients)
```

```{r tSNE broken}
detectVarGenes <- function (mat, ngenes = 500, FDR = 1e-05) 
{
  var.fit = scran::modelGeneVar(mat)
  var.out = scran::fitTrendVar(mat, var.fit)
  hvg.out = var.out[which(var.out$FDR <= FDR), ]
  hvg.out = hvg.out[order(hvg.out$bio, decreasing = TRUE),]
  genes = rownames(hvg.out)[1:ngenes]
  return(genes)
}
vg=detectVarGenes(suva_expr,500)
ts=calculateTsne(suva_expr,vg)
plotTsneGene(ts,suva_expr,c("MBP","CSF1R","ALDOC","OLIG1"))
```

```{r assign malig nonmalig}
lrbic=read.table(
  "~/Documents/Rubalind/Vignettes/SUVA_CNVs_BIC_LR.txt",sep="\t",header=T,row.names=1,check.names=F)
colnames(lrbic)
candRegions=rownames(lrbic)[which(lrbic[,"BIC difference"]>200 & lrbic[,"LRT adj. p-val"]<0.01)]
hi=plotHistogram(l[,candRegions],suva_expr,clusters=4,zscoreThreshold=4,patients)

normal= which(hi==1)
tumor=which(hi!=1)
```

```{r post probs w/ normal tumor cells}
redu=plotAll(suva_expr,normFactor,regions[candRegions,],gene_pos,"SUVA_CNVs_with_info.pdf",normal=normal,tumor=tumor)
```
```{r binary mtx}
bin_mat=binarizeMatrix(redu,normal,tumor,0.8)
plotBinaryMat(bin_mat,patients,normal,tumor,patient="MGH97")
```

```{r viz breakpoints}
plotAllChromosomes (mat = suva_expr,normal = normal,tumor = tumor,windowsize = 101,gene_pos = gene_pos,fname = "MGH36",patients = patients,patient = "MGH36",breakpoints = regions)
```







