
Analysis, visualization, and integration of spatial datasets with Seurat

https://satijalab.org/seurat/articles/spatial_vignette.html

# Overview

- analytical pipelines similar to Seurat workflow for scRNA-seq analysis
- updated interaction and visualization tools
- emphasis on integration of spatial and molecular information

Vignette:
- Normalization
- Dimensional reduction and clustering
- Detecting spatially-variable features
- Interactive visualization
- Integration with single-cell RNA-seq data
- Working with multiple slices

```{r Load libraries}
library(Seurat)
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
```

# 10x Visium data

- dataset of sagital mouse brain slices generated using the Visium v1 chemistry
- two serial anterior sections, and two (matched) serial posterior sections
- from output of `spaceranger` pipeline

```{r}
# InstallData("stxBrain")
?stxBrain
brain <- LoadData("stxBrain", type = "anterior1")
```

## Data preprocessing

- perform on spot by gene expression data
- normalize data to account for variance in seq depth across data points
variance in molecular counts / spot
- variance in molecular cts/spots can be substantial for spatial datasets
  - differences in cell density across tissue
- ex. data: substantial heterogeneity -> need effective normalization

```{r}
plot1 <- VlnPlot(brain, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
plot2 <- SpatialFeaturePlot(brain, features = "nCount_Spatial") + theme(legend.position = "right")
wrap_plots(plot1, plot2)
```

- variance in cts across spots not just technical, but dependent on tissue anatomy
- regions of tissue depleted for neurons reproducible exhibit lower molecular cts
- standard approaches (`LogNormalize()`) force data pts to have same underlying 'size' after normalization, problematic
- Satija `sctransform`: builds regularized neg bin models of gene expression
  - account for technical artifacts while preserving biological variance
  - normalizes data, detects high-variance features, stores data in `SCT` assay

```{r SCTransform}
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
```

## Gene expression viz

- overlay molecular data on top of tissue histology
- ex. data
  - `Hpca`: strong hippocampus marker
  - `Ttr`: choroid plexus

```{r SpatialFeaturePlot}
SpatialFeaturePlot(brain, features = c("Hpca", "Ttr"))
```
```{r save plot}
library(ggplot2)
plot <- SpatialFeaturePlot(brain, features = c("Ttr")) + theme(legend.text = element_text(size = 0),
    legend.title = element_text(size = 20), legend.key.size = unit(1, "cm"))
# jpeg(filename = "../output/images/spatial_vignette_ttr.jpg", height = 700, width = 1200, quality = 50)
print(plot)
# dev.off()
```

Adjust size of spots + transparency to improve histology viz
- `pt.size.factor`: scale size of spots, default is `1.6`
- `alpha`: min + max transparency, default is `c(1, 1)`
  - `alpha = c(0.1, 1)`: downweigh transparency of pts w/ lower expression

```{r}
p1 <- SpatialFeaturePlot(brain, features = "Ttr", pt.size.factor = 1)
p2 <- SpatialFeaturePlot(brain, features = "Ttr", alpha = c(0.1, 1))
p1 + p2
```

## Dimensionality reduction, clustering, viz

- same workflow as scRNAseq

```{r UMAP}
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:30)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:30)
```

```{r UMAP viz}
# No overlay
p1 <- DimPlot(brain, reduction = "umap", label = TRUE)
# Overlay
p2 <- SpatialDimPlot(brain, label = TRUE, label.size = 3)
p1 + p2
```
Make viz cleaner
```{r Viz adjustment}
SpatialDimPlot(brain, cells.highlight = CellsByIdentities(object = brain, idents = c(2, 1, 4, 3,
    5, 8)), facet.highlight = TRUE, ncol = 3)
```

```{r Interactive}
SpatialDimPlot(brain, interactive = TRUE)
SpatialFeaturePlot(brain, features = "Ttr", interactive = TRUE)
LinkedDimPlot(brain)
```

## Identification of spatially variable features

Seurat: 2 workflows
1. DE based on pre-annotated anatomical regions within tissue
  - determined from unsupervised clustering or prior knowledge
  - works in this case: clusters exhibit clear spatial restriction

```{r FindMarkers}
de_markers <- FindMarkers(brain, ident.1 = 5, ident.2 = 6)
SpatialFeaturePlot(object = brain, features = rownames(de_markers)[1:3], alpha = c(0.1, 1), ncol = 3)
```

2. `FindSpatiallyVariables()`
  - search for features exhibiting spatial patterning in absence of pre-annotation
  - default method `markvariogram` models data as mark point process, computes `variogram` which IDs genes whose expression level dependent on spatial location
    - inspired by Trendsceek
  - calculates gamma(r) values measuring dependence btwn 2 spots a certain `r` distance apart
  - default: use r-value of "5" in analyses, only compute values for variable genes
    - variation calculated independently of spatial location
  - other methods: `SpatialDE`, `Splotch`

```{r FindSpatiallyVariables}
brain <- FindSpatiallyVariableFeatures(brain, assay = "SCT", features = VariableFeatures(brain)[1:1000],
    selection.method = "moransi")

top.features <- head(SpatiallyVariableFeatures(brain, selection.method = "moransi"), 6)
SpatialFeaturePlot(brain, features = top.features, ncol = 3, alpha = c(0.1, 1))
```

## Subset out anatomical regions

- subset obj to focus on subset of data
- approx subset frontal cortex
- facilitate integration of data w/ cortical scRNAseq dataset in next section
- take subset of clusters, then segment based on exact positions

```{r Subset}
cortex <- subset(brain, idents = c(1, 2, 3, 4, 6, 7))
# now remove additional cells, use SpatialDimPlots to visualize what to remove
# SpatialDimPlot(cortex,cells.highlight = WhichCells(cortex, expression = image_imagerow > 400
# | image_imagecol < 150))
cortex <- subset(cortex, anterior1_imagerow > 400 | anterior1_imagecol < 150, invert = TRUE)
cortex <- subset(cortex, anterior1_imagerow > 275 & anterior1_imagecol > 370, invert = TRUE)
cortex <- subset(cortex, anterior1_imagerow > 250 & anterior1_imagecol > 440, invert = TRUE)
```

```{r Plot subset}
p1 <- SpatialDimPlot(cortex, crop = TRUE, label = TRUE)
p2 <- SpatialDimPlot(cortex, crop = FALSE, label = TRUE, pt.size.factor = 1, label.size = 3)
p1 + p2
```

## Integration w/ single-cell data
- ~50 um, Visium assay encompasses expression profiles of multiple cells
- deconvolute each spatial voxel to predict underlying comp of cell types
- superior performance using integration methods
  - substantially different noise models characterizing spatial + sc datasets
  - integration methods designed to be robust to these differences
- apply 'anchor'-based integration workflow from Seurat v3
  - probabilistic transfer of annotations from reference to query set
  
```{r Load data}
allen_reference <- readRDS("/Users/liaoyj/Downloads/allen_cortex.rds")
```

```{r Preprocess ref data}
# note that setting ncells=3000 normalizes the full dataset but learns noise models on 3k
# cells this speeds up SCTransform dramatically with no loss in performance
library(dplyr)
allen_reference <- SCTransform(allen_reference, ncells = 3000, verbose = FALSE) %>%
    RunPCA(verbose = FALSE) %>%
    RunUMAP(dims = 1:30)
```

```{r Renormalize cortex}
# After subsetting, we renormalize cortex
cortex <- SCTransform(cortex, assay = "Spatial", verbose = FALSE) %>%
    RunPCA(verbose = FALSE)
# the annotation is stored in the 'subclass' column of object metadata
DimPlot(allen_reference, group.by = "subclass", label = TRUE)
```

```{r Label transfer}
anchors <- FindTransferAnchors(reference = allen_reference, query = cortex, normalization.method = "SCT")
predictions.assay <- TransferData(anchorset = anchors, refdata = allen_reference$subclass, prediction.assay = TRUE,
    weight.reduction = cortex[["pca"]], dims = 1:30)
cortex[["predictions"]] <- predictions.assay
```





